packages/flutter_tools/lib/src/migrate/migrate_compute.dart

```diff
@@ -0,0 +1,620 @@
+    required this.flutterProject,
+    required this.blacklistPrefixes,
+    required this.logger,
+    required this.verbose,
+    required this.fileSystem,
+    required this.status,
+    required this.migrateUtils,
+    this.baseProject,
+    this.targetProject,
+  });
+
+  MigrateResult migrateResult;
+  FlutterProject flutterProject;
+  Set<String?> blacklistPrefixes;
+  Logger logger;
+  bool verbose;
+  FileSystem fileSystem;
+  Status status;
+  MigrateUtils migrateUtils;
+
+  MigrateBaseFlutterProject? baseProject;
+  MigrateTargetFlutterProject? targetProject;
+}
+
+String getLocalPath(String path, String basePath, FileSystem fileSystem) {
+  return path.replaceFirst(basePath + fileSystem.path.separator, '');
+}
+
+/// Returns a base revision to fallback to in case a true base revision is unknown.
+String _getFallbackBaseRevision(bool allowFallbackBaseRevision, bool verbose, Logger logger, Status status) {
+  if (!allowFallbackBaseRevision) {
+    status.stop();
+    logger.printError('Could not determine base revision this app was created with:');
+    logger.printError('.metadata file did not exist or did not contain a valid revision.', indent: 2);
+    logger.printError('Run this command again with the `--allow-fallback-base-revision` flag to use Flutter v1.0.0 as the base revision or manually pass a revision with `--base-revision=<revision>`', indent: 2);
+    throwToolExit('Failed to resolve base revision');
+  }
+  // Earliest version of flutter with .metadata: c17099f474675d8066fec6984c242d8b409ae985 (2017)
+  // Flutter 2.0.0: 60bd88df915880d23877bfc1602e8ddcf4c4dd2a
+  // Flutter v1.0.0: 5391447fae6209bb21a89e6a5a6583cac1af9b4b
+  //
+  // TODO(garyq): Use things like dart sdk version and other hints to better fine-tune this fallback.
+  //
+  // We fall back on flutter v1.0.0 if .metadata doesn't exist.
+  if (verbose) {
+    status.pause();
+    logger.printStatus('Could not determine base revision, falling back on `v1.0.0`, revision 5391447fae6209bb21a89e6a5a6583cac1af9b4b', color: TerminalColor.grey, indent: 4);
+    status.resume();
+  }
+  return '5391447fae6209bb21a89e6a5a6583cac1af9b4b';
+}
+
+abstract class MigrateFlutterProject {
+  MigrateFlutterProject({
+    required this.path,
+    required this.directory,
+    required this.name,
+    required this.androidLanguage,
+    required this.iosLanguage,
+    this.platformWhitelist,
+  });
+
+  final String? path;
+  final Directory directory;
+  final String name;
+  final String androidLanguage;
+  final String iosLanguage;
+  final List<SupportedPlatform>? platformWhitelist;
+
+  /// Run git diff over each matching pair of files in the this project and the provided target project.
+  Future<Map<String, DiffResult>> diff(
+    MigrateContext context,
+    MigrateFlutterProject other,
+  ) async {
+    final Map<String, DiffResult> diffMap = <String, DiffResult>{};
+    final List<FileSystemEntity> thisFiles = directory.listSync(recursive: true);
```

Comment:
I'm not sure if this is right, but should we instead be parsing `git ls-files`, which would skip any patterns matching .gitignore?

Here's an example where we do this in the analysis shard: https://github.com/flutter/flutter/blob/master/dev/bots/analyze.dart#L1521

Reply:
Oh, nvm, I see you're doing this in `migrateUtils.isGitIgnored`, nice.

Reply:
Yeah, git ls-files would be better, but either way, I check for gitignored files later on, so it should all be filtered out.